src/uct/base/uct_md.c

```diff
@@ -368,22 +368,94 @@ static ucs_status_t uct_mem_check_flags(unsigned flags)
     return UCS_OK;
 }
 
-ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *alloc_name, uct_mem_h *memh_p)
+ucs_status_t uct_mem_alloc_check_params(const uct_mem_alloc_params_t *params)
 {
+    const uct_alloc_method_t *method;
     ucs_status_t status;
 
-    status = uct_mem_check_flags(flags);
+    if (!(params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_FLAGS)) {
+        ucs_debug("UCT_MEM_ALLOC_PARAM_FIELD_FLAGS not set");
+    } else {
+        status = uct_mem_check_flags(params->flags);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    if (!(params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_LENGTH)) {
+        ucs_debug("the length value for allocating memory isn't set: %s",
+                  ucs_status_string(UCS_ERR_INVALID_PARAM));
+        return UCS_ERR_INVALID_PARAM;
+    } else if (params->length == 0) {
+        ucs_debug("the length value for allocating memory is set to zero: %s",
+                  ucs_status_string(UCS_ERR_INVALID_PARAM));
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if ((params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_FLAGS) &&
+        (params->flags & UCT_MD_MEM_FLAG_FIXED)) {
+        if (!(params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_ADDR)) {
+            ucs_debug("UCT_MD_MEM_FLAG_FIXED requires setting of"
+                      " UCT_MEM_ALLOC_PARAM_FIELD_ADDR field");
+            return UCS_ERR_INVALID_PARAM;
+        } else if (!params->address ||
+                   ((uintptr_t)params->address % ucs_get_page_size())) {
+            ucs_debug("UCT_MD_MEM_FLAG_FIXED requires valid page size aligned address");
+            return UCS_ERR_INVALID_PARAM;
+        }
+	}
+
+    if (!(params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_METHODS)) {
+        ucs_debug("allocation methods field not set");
+        return UCS_ERR_INVALID_PARAM;
+    } else if (params->methods.count == 0) {
+        ucs_debug("count of allocation methods provided is set to zero: %s",
+                  ucs_status_string(UCS_ERR_INVALID_PARAM));
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    for (method = params->methods.methods;
+         method < (params->methods.methods + params->methods.count);
+         ++method) {
+
+        if (*method == UCT_ALLOC_METHOD_MD) {
+
+            if (params->mds.count == 0) {
+                ucs_debug("count of mds provided is set to zero: %s",
+                          ucs_status_string(UCS_ERR_INVALID_PARAM));
+                return UCS_ERR_INVALID_PARAM;
+            }
+        }
+
+        if (((*method == UCT_ALLOC_METHOD_MMAP) ||
+             (*method == UCT_ALLOC_METHOD_THP) ||
+             (*method == UCT_ALLOC_METHOD_HEAP) ||
+             (*method == UCT_ALLOC_METHOD_HUGE)) &&
+            (!(params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_ADDR))) {
+            ucs_debug("address field must be set when mmap method is used");
```

Comment:
why we need address to allocate?

Reply:
Otherwise we would need to initialize param->address to NULL but uct_mem_alloc_params_t is `const` type.

Reply:
not really, when user allocates memory they can avoid passing address and allow the allocation to generate the address. we should pass NULL to mmap internally(), not via params struct